web/source/osk/embedded/globeHint.ts

```diff
@@ -0,0 +1,164 @@
+        const _Box = this.vkbd.element.parentNode as HTMLDivElement;
+        const _BoxRect = _Box.getBoundingClientRect();
+        const keyRect = key.getBoundingClientRect();
+        let y = (keyRect.bottom - _BoxRect.top + 1);
+
+        // Width dimensions must be set explicitly to prevent clipping.
+        // We'll assume that the globe key is always positioned on the bottom row.
+        let oskBaseFontSize = new ParsedLengthStyle(keyman.osk.baseFontSize);
+        let bubbleWidth = Math.ceil(xWidth * 3);
+
+        // The base OSK font-size is typically set in 'em', and we want to scale up
+        // the bubble accordingly for devices where it'd display too small otherwise.
+        if(!oskBaseFontSize.absolute) {
+          bubbleWidth *= oskBaseFontSize.val;
+        }
+
+        bubbleStyle.bottom = Math.floor(keyman.osk.computedHeight - y) + 'px';
+        // CSS already defines transform: translateX(-50%) - this centers the element.
+        bubbleStyle.left = center + 'px';
+
+        this.tip.style.bottom = (rrow.height - 1) + 'px';
+        this.tip.style.width = bubbleWidth + 'px';
+
+        // Adjust shape if at edges
+
+        // how much width lies outside the range of the base key, per side
+        var xOverflow = (bubbleWidth - xWidth) / 2;
+
+        let capWidth = xWidth * 2 / 3;
+
+        const leftEdgeMargin  = xLeft;
+        const rightEdgeMargin = window.innerWidth - (xLeft + xWidth);
+
+        if(leftEdgeMargin < xOverflow) {
+          // The overflow would be clipped by the left edge when centered,
+          // so we "offset" this part to keep it within screen bounds.
+          this.tip.style.transform = 'translateX(' + (xOverflow - leftEdgeMargin + 1) + 'px)';
+        } else if(rightEdgeMargin < xOverflow) {
+          // The overflow would be clipped by the right edge when centered,
+          // so we "offset" this part to keep it within screen bounds.
+          this.tip.style.transform = 'translateX(-' + (xOverflow - rightEdgeMargin + 1) + 'px)';
+        }
+
+        if(!oskBaseFontSize.absolute) {
+          this.element.style.fontSize = keyman.osk.baseFontSize;
+        }
+
+        let capHeight = xHeight / 3;
+
+        this.cap.style.bottom = (rrow.height - capHeight) + 'px';
+        this.cap.style.width = '0px';
+        this.cap.style.height = '0px';
+        this.cap.style.borderLeftWidth   = (capWidth / 2) + 'px';
+        this.cap.style.borderRightWidth  = (capWidth / 2) + 'px';
+        this.cap.style.borderTopWidth    = capHeight + 'px';
+
+        bubbleStyle.display = 'block';
+      }
+
+      // Save the key preview state
+      this.state = on;
+    }
+
+    public show(key: KeyElement, onDismiss?: () => void) {
+      this._show(key, true);
+
+      if(onDismiss) {
+        // enable input-capture system
+        this.inputTrap = new SingleInputCapture(() => {
+          this._show(key, false);
+          onDismiss();
```

Comment:
I think we want to set `this.inputTrap = null;` in the callback. And probably also in the `else` for `if(onDismiss)`

Reply:
I think we go for consistency and try to avoid side-effects.